Enable phase transitions in multicomponent compressible

naliboff opened the pull request:
This pull request enables phase transitions in `equation_of_state/multicomponent_compressible` and `material_model/multiple_compressible`.


* [X] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

jdannberg commented on tests/multicomponent_compressible_averaging_phases.prm:
```diff
@@ -0,0 +1,42 @@
+# This test checks that the addition of phase transitions with corresponding
+# change material properties works correctly in the multicomponent material
```
```suggestion
# changes in the material properties works correctly in the multicomponent material
```

jdannberg commented on tests/multicomponent_compressible_averaging_phases.prm:
```diff
@@ -0,0 +1,42 @@
+
+
+# Provide a name for the compositional field, which will be
+# used to define change in material properties across phase
```
```suggestion
# used to define the change in material properties across phase
```

jdannberg commented on tests/multicomponent_compressible_averaging_phases.prm:
```diff
@@ -0,0 +1,42 @@
+
+  subsection Multicomponent compressible
+
+    # Add a single phase transition approximately the model mid-depth,
```
```suggestion
    # Add a single phase transition at approximately the model mid-depth,
```

jdannberg commented on tests/multicomponent_compressible_averaging_phases.prm:
```diff
@@ -0,0 +1,42 @@
+  subsection Multicomponent compressible
+
+    # Add a single phase transition approximately the model mid-depth,
+    # but do not change any of the material properties.
```
But you are changing the material properties below.

jdannberg commented on tests/multicomponent_compressible_averaging_phases.prm:
```diff
@@ -0,0 +1,42 @@
+# model. The test is a modification of tests/multicomponent_compressible_averaging.prm,
+# with the primary changes including the naming of a compositional field, a 
+# phase transition, and corresponding changes in material properties.
+include $ASPECT_SOURCE_DIR/tests/multicomponent_compressible_averaging.prm
```
The multicomponent_compressible_averaging test has a very nice description of what output to expect (i.e., what's the expected result of the averaging). I think you changed the parameters in the lower half of the box, so would it make sense to write a short description here about what material numbers to expect in the output, just so that we know the test results are correct?

jdannberg reviewed (commented):
I general, this looks good to me (other than a few comments). But I'd really like to hear from @bobmyhill about what this does to the thermodynamic consistency of this material model.

naliboff commented on .vscode/launch.json:
```diff
@@ -1,38 +1,20 @@
 {
```
Oh dear, I'm not sure how that happened.

jdannberg commented:
Oh, also: Can you add a changelog entry?

naliboff commented on source/material_model/equation_of_state/multicomponent_compressible.cc:
```diff
@@ -75,37 +75,37 @@ namespace aspect
-                           "for a total of N+1 values, where N is the number of compositional fields."
+                           Patterns::Anything(),
+                           "List of reference temperatures $T_0$ for background and compositional fields,"
+                           "for a total of N+P+1 values, where N is the number of compositional fields and P is the number of phases. "
```
@jdannberg - How does the following sound?

```
"List of reference temperatures $T_0$ for background and compositional fields (N), 
"for a total of N+1 values for models with no phase transitions. Models with phase "
"transitions (P) contain a list of field names (N+1) and P+1 values specified for each field, "
"If only one value is given, then all use the same value. "
```

naliboff commented on source/material_model/multicomponent_compressible.cc:
```diff
@@ -60,10 +96,33 @@ namespace aspect
           // This may not be the most reasonable thing, but for most Earth materials we hope
           // that they do not vary so much that it is a big problem.
-          out.thermal_conductivities[i] = MaterialUtilities::average_value (volume_fractions, thermal_conductivities, MaterialUtilities::arithmetic);
+          std::vector<double> phase_averaged_thermal_conductivities(volume_fractions.size());
+          for (unsigned int c=0; c<volume_fractions.size(); ++c)
+            phase_averaged_thermal_conductivities[c] = MaterialUtilities::phase_average_value(phase_function_values,
+                                                       n_phase_transitions_for_each_chemical_composition,
+                                                       thermal_conductivities,
+                                                       c);
+          out.thermal_conductivities[i] = MaterialUtilities::average_value (volume_fractions, phase_averaged_thermal_conductivities, MaterialUtilities::arithmetic);
```
@jdannberg - good catch. It is parsed correctly, but the declaration does need to be updated.

naliboff commented on source/material_model/multicomponent_compressible.cc:
```diff
@@ -60,10 +96,33 @@ namespace aspect
 
           // User-defined volume fraction averaging of viscosities
-          out.viscosities[i] = MaterialUtilities::average_value (volume_fractions, viscosities, viscosity_averaging);
+          if (in.requests_property(MaterialProperties::viscosity) || in.requests_property(MaterialProperties::additional_outputs) ||
+              (this->get_parameters().enable_elasticity && in.requests_property(MaterialProperties::reaction_rates) ) ||
+              (this->get_adiabatic_conditions().is_initialized()))
```
I had originally thought there was a bug in the code related to the viscosity averaging when the initial conditions were initialized. These checks were put in to try and resolve the bug, but it turns out it was related to the thermal conductivity averaging. However, I think we should leave the other checks in, as it should improve performance and follows the structure implemented in other material models.

naliboff commented:
@jdannberg - Thanks for the review! I've addressed most of the comments, including replacing one of the tests with a box geometry test case where it is much easier to verify the phase transitions are behaving correctly for a case with multiple compositional fields experiencing a different number of phase transitions. At this stage, maybe we should wait for @bobmyhill to comment?

jdannberg commented on include/aspect/geometry_model/spherical_shell.h:
```diff
@@ -153,13 +153,12 @@ namespace aspect
           double topography_for_point (const Point<dim> &x_y_z) const;
 
+        private:
           /**
            * A pointer to the topography model.
            */
           const InitialTopographyModel::Interface<dim> *topo;
 
-        private:
-
```
I am confused. This has nothing to do with the rest of this PR, right?

jdannberg commented on include/aspect/material_model/equation_of_state/multicomponent_compressible.h:
```diff
@@ -34,13 +34,18 @@ namespace aspect
+       * with phase transitions, the list needs to contain each field name, including the background,
+       * for a total of N+1 names, and for each of these names, specify the value for each phase
+       * Therefore, the total number of values given is N+P+1, with P = sum(P_c) the total number of
+       * phase transitions, summed over all phases. The format is background: value1|value2|...|valueP_1+1,
+       * field1:value1|...|valueP_2+1, ..., fieldN: value1|...|valueP_N+1. If only one value is given,
+       * then all fields/phases use the same value. Other lengths of lists are not allowed.
```
I think this part is not needed here (since a user would not look for it in the .h file). 
```suggestion
       * phase transitions, summed over all phases. 
```

jdannberg commented on include/aspect/material_model/equation_of_state/multicomponent_compressible.h:
```diff
@@ -34,13 +34,18 @@ namespace aspect
+       * field1:value1|...|valueP_2+1, ..., fieldN: value1|...|valueP_N+1. If only one value is given,
+       * then all fields/phases use the same value. Other lengths of lists are not allowed.
+       *
+       * The material parameters for each compositional field and phase are calculated self-consistently,
```
Is this where you wanted to add a comment about the phase transitions breaking this consistency?

jdannberg commented on source/geometry_model/spherical_shell.cc:
```diff
@@ -713,7 +713,7 @@ namespace aspect
     SphericalShell<dim>::maximal_depth() const
     {
-      return R1 + manifold->topo->max_topography() - R0;
+      return R1 + this->get_initial_topography_model().max_topography() - R0;
```
Same here: why is that change here?

jdannberg commented:
@bobmyhill 
We had a discussion this morning, and the solution we came up with is to add an input parameter along the lines of "Enable phase transitions" that is by default set to false and documents that switching it to true breaks thermodynamic consistency and why. If we later on want to add the proper functionality we could still implement Equations 20, 21, 23 from Stixrude and Lithgow-Bertelloni.
@naliboff does not actually need the phase transitions, only the infrastructure (to include it in viscoplastic without making the logic horribly complicated), so we also debated simply putting in an Assert that disables it, but that means that we also could not test the code, so we decided against that. 

Does this sound like a reasonable solution?

bobmyhill commented:
That solution sounds like the best option, I'm happy for us to choose this.

naliboff commented on source/material_model/equation_of_state/multicomponent_compressible.cc:
```diff
@@ -75,52 +75,113 @@ namespace aspect
-                           "for a total of N+1 values, where N is the number of compositional fields."
-                           "If only one value is given, then all use the same value. "
+                           Patterns::Anything(),
+                           "List of reference densities $T_0$ for background and compositional fields (N), "
```
@jdannberg - Would you like this change applied to all of the parameter documentations? I think if we change it here it should be consistent across all parameters, but in many cases the background reference density may not be associated with the mantle (ex: a crustal-scale model), so my preference would be to keep the description more general.

naliboff commented:
@jdannberg - I think this is ready for another look when you are ready. 

@bobmyhill - I forgot to say thank you earlier for the explanation of why the phase transitions would break thermodynamic consistency and suggestions for how to proceed!

jdannberg commented on include/aspect/material_model/equation_of_state/multicomponent_compressible.h:
```diff
@@ -34,18 +34,22 @@ namespace aspect
+       * with phase transitions, the list needs to contain each field name, including the background,
+       * for a total of N+1 names, and for each of these names, specify the value for each phase
+       * Therefore, the total number of values given is N+P+1, with P = sum(P_c) the total number of
+       * phase transitions, summed over all phases. The format is background: value1|value2|...|valueP_1+1,
+       * field1:value1|...|valueP_2+1, ..., fieldN: value1|...|valueP_N+1. If only one value is given,
+       * then all fields/phases use the same value. Other lengths of lists are not allowed.
```
```suggestion
       * phase transitions, summed over all phases. 
```

Again, I would remove this. It will just get outdated if we change the interface. and nobody will look for it in this place. 
It's not the responsibility of the documentation in the .h file to describe the format of parameters in the input file.

jdannberg commented on include/aspect/material_model/equation_of_state/multicomponent_compressible.h:
```diff
@@ -56,15 +60,19 @@ namespace aspect
        * $f$ is a scaling factor for $\alpha$ and $\beta_T$.
+       *
+       * Significantly, if phase transitions are included the formulation no longer self-consistently calculates
+       * the second derivative properties (heat capacity, thermal expansivity, and compressibility), as they are
+       * they are affected by the P-T-X dependence of the phase function.
```
```suggestion
       * the second derivative properties (heat capacity, thermal expansivity, and compressibility), as they are
       * affected by the P-T-X dependence of the phase function. In terms of physics, this means that latent heat
       * and excess expansivity/compressibility of reactions are neglected. Fixing that would require implementing
       * equations 20, 21, 23 in Stixrude and Lithgow-Bertelloni, 2022 (https://doi.org/10.1093/gji/ggab394).
```

jdannberg commented on source/material_model/equation_of_state/multicomponent_compressible.cc:
```diff
@@ -83,56 +83,132 @@ namespace aspect
+                           "Whether to enable the use of phase transitions, which break the thermodynamic "
+                           "consistency of the equation of state for properties (heat capacity, thermal "
+                           "expansivity and compressibility) that are affected by the P-T-X dependence "
+                           "of the phase transition.");
```
```suggestion
                           "of the phase transition. Specifically, this means that latent heat "
                           "and excess expansivity/compressibility of reactions are neglected.");
```

jdannberg commented on source/material_model/equation_of_state/multicomponent_compressible.cc:
```diff
@@ -147,6 +223,24 @@ namespace aspect
+          {
+            for (unsigned int n=0; n < (*expected_n_phases_per_composition).size(); ++n)
+              {
+                AssertThrow((*expected_n_phases_per_composition)[n] == 1, ExcMessage(
+                              "The number of expected phases per composition for the compositional "
```
Can you make that more readable?
```suggestion
                AssertThrow((*expected_n_phases_per_composition)[n] == 1, 
                              ExcMessage("The number of expected phases per composition for the compositional "
```

jdannberg commented on source/material_model/equation_of_state/multicomponent_compressible.cc:
```diff
@@ -83,56 +83,132 @@ namespace aspect
-                           "for a total of N+1 values, where N is the number of compositional fields."
-                           "If only one value is given, then all use the same value. "
+                           Patterns::Anything(),
+                           "List of reference densities $T_0$ for background and compositional fields (N), "
```
```suggestion
                           "List of reference densities for background and compositional fields (N), "
```

jdannberg reviewed (commented):
I still have a few small comments on formatting and documentation, but other wise this looks good to me! Thank you!

jdannberg commented:
Oh, and there's something wrong with your new algorithm of reading in the maps in parse_parameters()

Several tests crash with
```
----------------------------------------------------
Exception 'ExcMessage("parse_map_to_double_array can only check the structure " "of the parsed map for " + options.property_name + " if an expected number of values for each key is given. " "The expected number of values is " + std::to_string(options.list_of_required_keys.size()) + ", but instead " + std::to_string(options.n_values_per_key.size()) + " were provided.")' on rank 0 on processing: 

--------------------------------------------------------
An error occurred in line <373> of file </__w/aspect/aspect/source/utilities.cc> in function
    std::vector<double> aspect::Utilities::MapParsing::parse_map_to_double_array(const string&, aspect::Utilities::MapParsing::Options&)
The violated condition was: 
    options.check_values_per_key == false || options.n_values_per_key.size() == options.list_of_required_keys.size()
Additional information: 
    parse_map_to_double_array can only check the structure of the parsed
    map for Reference temperatures if an expected number of values for
    each key is given. The expected number of values is 2, but instead 1
    were provided.
```
and they have totally standard input parameters:
```
subsection Material model
  set Model name = multicomponent compressible

  subsection Multicomponent compressible
    set Reference temperatures                       = 1600
    set Reference densities                          = 3300
    set Reference isothermal compressibilities       = 4e-12
    set Isothermal bulk modulus pressure derivatives = 4
    set Reference thermal expansivities              = 2.e-5
    set Isochoric specific heats                     = 1250
    set Thermal conductivities                       = 0.0
  end
end

```

naliboff commented:
@jdannberg - Thank you for taking another look and the comments! I started working through the error you noted in the last comment this afternoon, and I think I fixed the bug in the last commit (at least the tests worked locally, we'll see what the testers say). If that fixes things, I'll work on addressing your comments tomorrow morning and hopefully this will be ready to go.
